test/jb_controller/burn_tokens_of.test.js

```diff
@@ -0,0 +1,343 @@
+      .returns(false);
+
+    await mockJbOperatorStore.mock.hasPermission
+      .withArgs(terminalSigner.address, projectOwner.address, 0, BURN_INDEX)
+      .returns(false);
+
+    await mockJbDirectory.mock.isTerminalDelegateOf
+      .withArgs(PROJECT_ID, terminalSigner.address)
+      .returns(true);
+
+    await expect(
+      jbController
+        .connect(terminalSigner)
+        .burnTokensOf(
+          holder.address,
+          PROJECT_ID,
+          AMOUNT_TO_BURN,
+          MEMO,
+          /*_preferClaimedTokens=*/ true,
+        ),
+    ).to.be.not.reverted;
+  });
+
+  it(`Can't burn 0 token`, async function () {
+    const { holder, jbController } = await setup();
+
+    await expect(
+      jbController
+        .connect(holder)
+        .burnTokensOf(
+          holder.address,
+          PROJECT_ID,
+          /*_tokenCount=*/ 0,
+          MEMO,
+          /*_preferClaimedTokens=*/ true,
+        ),
+    ).to.be.revertedWith(errors.NO_BURNABLE_TOKENS);
+  });
+
+  it(`Can't burn token if funding cycle is paused and caller is not a terminal delegate`, async function () {
+    const { holder, jbController, mockJbFundingCycleStore, timestamp } = await setup();
+
+    await mockJbFundingCycleStore.mock.currentOf.withArgs(PROJECT_ID).returns({
+      // mock JBFundingCycle obj
+      number: 1,
+      configuration: timestamp,
+      basedOn: timestamp,
+      start: timestamp,
+      duration: 0,
+      weight: 0,
+      discountRate: 0,
+      ballot: ethers.constants.AddressZero,
+      metadata: packFundingCycleMetadata({ pauseBurn: 1, reservedRate: RESERVED_RATE }),
+    });
+
+    await expect(
+      jbController
+        .connect(holder)
+        .burnTokensOf(
+          holder.address,
+          PROJECT_ID,
+          AMOUNT_TO_BURN,
+          MEMO,
+          /*_preferClaimedTokens=*/ true,
+        ),
+    ).to.be.revertedWith(errors.BURN_PAUSED_AND_SENDER_NOT_VALID_TERMINAL_DELEGATE);
+  });
+
+  it(`Should burn token if funding cycle is paused and caller is a terminal delegate`, async function () {
+    const {
+      projectOwner,
+      holder,
+      jbController,
+      mockJbFundingCycleStore,
+      mockJbOperatorStore,
+      mockJbDirectory,
+      timestamp,
+    } = await setup();
+    const terminal = await deployMockContract(projectOwner, jbTerminal.abi);
+    const terminalSigner = await impersonateAccount(terminal.address);
+
+    await mockJbOperatorStore.mock.hasPermission
+      .withArgs(terminalSigner.address, projectOwner.address, PROJECT_ID, BURN_INDEX)
+      .returns(false);
+
+    await mockJbOperatorStore.mock.hasPermission
+      .withArgs(terminalSigner.address, projectOwner.address, 0, BURN_INDEX)
+      .returns(false);
+
+    await mockJbDirectory.mock.isTerminalDelegateOf
+      .withArgs(PROJECT_ID, terminalSigner.address)
+      .returns(true);
+
+    await mockJbFundingCycleStore.mock.currentOf.withArgs(PROJECT_ID).returns({
+      // mock JBFundingCycle obj
```

Comment:
I'm not sure these comments are needed everywhere. the code is self-documenting to me.